x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/autoscaling/MlAutoscalingResourceTracker.java

```diff
@@ -230,9 +272,86 @@ static void getMemoryAndCpu(
+    static boolean tryRemoveOneNode(Map<String, List<Long>> perNodeUsedModelMemoryInBytes, long perNodeMemoryInBytes) {
+        if (perNodeUsedModelMemoryInBytes.size() <= 1) {
+            return false;
+        }
+
+        Map<String, Long> perNodeModelMemoryInBytesSum = perNodeUsedModelMemoryInBytes.entrySet()
+            .stream()
+            .map(entry -> tuple(entry.getKey(), entry.getValue().stream().mapToLong(Long::longValue).sum()))
+            .collect(Collectors.toMap(Tuple::v1, Tuple::v2));
+
+        Optional<Map.Entry<String, Long>> leastLoadedNodeAndMemoryUsage = perNodeModelMemoryInBytesSum.entrySet()
+            .stream()
+            .min(Comparator.comparingLong(Map.Entry::getValue));
+
+        if (leastLoadedNodeAndMemoryUsage.isPresent() == false) {
+            return false;
+        }
+
+        // this currently only works based on memory, not CPU
+        String candidateNode = leastLoadedNodeAndMemoryUsage.get().getKey();
+        List<Long> candidateNodeMemoryList = perNodeUsedModelMemoryInBytes.get(candidateNode);
+        perNodeModelMemoryInBytesSum.remove(candidateNode);
+
+        // if all jobs fit on other nodes, we can scale down one node
+        return tryMoveJobsByMemoryInLeastEfficientWay(candidateNodeMemoryList, perNodeModelMemoryInBytesSum, perNodeMemoryInBytes) == 0L;
+    }
+
+    /**
+     * Try to place jobs by size to other nodes in the leas efficient way
+     *
+     * @param candidateNodeMemoryList list of jobmemory values given by the candidate job
+     * @param perNodeModelMemoryInBytesSum other nodes memory usage
+     * @param perNodeMemoryInBytes available memory per node
+     * @return remaining memory, that could not be placed on other nodes, 0L if all jobs got placed
+     */
+    static long tryMoveJobsByMemoryInLeastEfficientWay(
+        List<Long> candidateNodeMemoryList,
+        Map<String, Long> perNodeModelMemoryInBytesSum,
+        long perNodeMemoryInBytes
+    ) {
+        if (candidateNodeMemoryList.size() == 0) {
+            return 0L;
+        }
+
+        // sort ascending so we put the smaller jobs first, by intention this might make it impossible to place the bigger jobs at the end
+        List<Long> candidateNodeMemoryListSorted = candidateNodeMemoryList.stream().sorted().toList();
+        long candidateNodeMemorySum = candidateNodeMemoryList.stream().mapToLong(Long::longValue).sum();
+
+        if (perNodeModelMemoryInBytesSum.size() == 0) {
+            return candidateNodeMemorySum;
+        }
+
+        // use a priority queue that keeps the nodes with the lowest memory usage on top(all have the same size)
+        // that way we assign the jobs to the node with most available memory first
+        PriorityQueue<Long> modelMemoryInBytesSums = new PriorityQueue<>(perNodeModelMemoryInBytesSum.values());
+
+        for (Long jobMemory : candidateNodeMemoryListSorted) {
+            Long memoryAfterAddingJobMemory = modelMemoryInBytesSums.poll() + jobMemory;
```

Comment:
nit: The IDE is telling me that `memoryAfterAddingJobMemory` can be a primitive `long`, I don't know if that prevents autoboxing in someway 🤷‍♂️ .

nit: Is it worth checking for null for the `modelMemoryInBytesSums.poll()` call? It doesn't look like it could ever be null since we're checking for `size() == 0` above and adding the value that is removed anyway.